libsolidity/ast/TypeProvider.cpp

```diff
@@ -0,0 +1,430 @@
+{
+	auto& map = _modifier == FixedPointType::Modifier::Unsigned ? get().m_ufixedMxN : get().m_fixedMxN;
+
+	auto i = map.find(make_pair(m, n));
+	if (i != map.end())
+		return i->second.get();
+
+	return map.emplace(
+		make_pair(m, n),
+		make_unique<FixedPointType>(
+			m,
+			n,
+			_modifier
+		)
+	).first->second.get();
+}
+
+TupleType const* TypeProvider::tupleType(vector<Type const*> members)
+{
+	if (members.empty())
+		return &m_emptyTupleType;
+
+	return createAndGet<TupleType>(move(members));
+}
+
+ReferenceType const* TypeProvider::withLocation(ReferenceType const* _type, DataLocation _location, bool _isPointer)
+{
+	if (_type->location() == _location && _type->isPointer() == _isPointer)
+		return _type;
+
+	get().m_generalTypes.emplace_back(_type->copyForLocation(_location, _isPointer));
+	return static_cast<ReferenceType const*>(get().m_generalTypes.back().get());
+}
+
+FunctionType const* TypeProvider::functionType(FunctionDefinition const& _function, bool _isInternal)
+{
+	return createAndGet<FunctionType>(_function, _isInternal);
+}
+
+FunctionType const* TypeProvider::functionType(VariableDeclaration const& _varDecl)
+{
+	return createAndGet<FunctionType>(_varDecl);
+}
+
+FunctionType const* TypeProvider::functionType(EventDefinition const& _def)
+{
+	return createAndGet<FunctionType>(_def);
+}
+
+FunctionType const* TypeProvider::functionType(FunctionTypeName const& _typeName)
+{
+	return createAndGet<FunctionType>(_typeName);
+}
+
+FunctionType const* TypeProvider::functionType(
+	strings const& _parameterTypes,
+	strings const& _returnParameterTypes,
+	FunctionType::Kind _kind,
+	bool _arbitraryParameters,
+	StateMutability _stateMutability
+)
+{
+	return createAndGet<FunctionType>(
+		_parameterTypes, _returnParameterTypes,
+		_kind, _arbitraryParameters, _stateMutability
+	);
+}
+
+FunctionType const* TypeProvider::functionType(
+	TypePointers const& _parameterTypes,
+	TypePointers const& _returnParameterTypes,
+	strings _parameterNames,
+	strings _returnParameterNames,
+	FunctionType::Kind _kind,
+	bool _arbitraryParameters,
+	StateMutability _stateMutability,
+	Declaration const* _declaration,
+	bool _gasSet,
+	bool _valueSet,
+	bool _bound
+)
+{
+	return createAndGet<FunctionType>(
+		_parameterTypes,
+		_returnParameterTypes,
+		_parameterNames,
+		_returnParameterNames,
+		_kind,
+		_arbitraryParameters,
+		_stateMutability,
+		_declaration,
+		_gasSet,
+		_valueSet,
+		_bound
+	);
+}
+
+RationalNumberType const* TypeProvider::rationalNumberType(rational const& _value, Type const* _compatibleBytesType)
+{
+	return createAndGet<RationalNumberType>(_value, _compatibleBytesType);
+}
+
+ArrayType const* TypeProvider::arrayType(DataLocation _location, bool _isString)
+{
+	if (_isString)
+	{
+		if (_location == DataLocation::Storage)
+			return stringType();
+		if (_location == DataLocation::Memory)
+			return stringMemoryType();
```

Comment:
What about calldata?

Reply:
it's just an optimization. the catch-all is below. (as discussed just right now)